Index: material-components-android-master/lib/java/com/google/android/material/AndroidManifest.xml
===================================================================
--- material-components-android-master.orig/lib/java/com/google/android/material/AndroidManifest.xml
+++ /dev/null
@@ -1,20 +0,0 @@
-<?xml version="1.0" encoding="utf-8"?>
-<!-- Copyright (C) 2015 The Android Open Source Project
-
-     Licensed under the Apache License, Version 2.0 (the "License");
-     you may not use this file except in compliance with the License.
-     You may obtain a copy of the License at
-
-          http://www.apache.org/licenses/LICENSE-2.0
-
-     Unless required by applicable law or agreed to in writing, software
-     distributed under the License is distributed on an "AS IS" BASIS,
-     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-     See the License for the specific language governing permissions and
-     limitations under the License.
--->
-<manifest xmlns:android="http://schemas.android.com/apk/res/android"
-    package="com.google.android.material">
-
-  <application/>
-</manifest>
Index: material-components-android-master/lib/java/com/google/android/material/badge/BadgeDrawable.java
===================================================================
--- material-components-android-master.orig/lib/java/com/google/android/material/badge/BadgeDrawable.java
+++ material-components-android-master/lib/java/com/google/android/material/badge/BadgeDrawable.java
@@ -490,6 +490,17 @@ public class BadgeDrawable extends Drawa
     }
   }
 
+  public boolean isBadgeNotSaveColor() {
+    return state.isNotSaveColor();
+  }
+
+  public void setBadgeNotSaveColor(boolean isNotSaveColor) {
+    if (isNotSaveColor != state.isNotSaveColor()) {
+      state.setNotSaveColor(isNotSaveColor);
+      invalidateSelf();
+    }
+  }
+
   /** Returns whether this badge will display a number. */
   public boolean hasNumber() {
     return !state.hasText() && state.hasNumber();
Index: material-components-android-master/lib/java/com/google/android/material/badge/BadgeState.java
===================================================================
--- material-components-android-master.orig/lib/java/com/google/android/material/badge/BadgeState.java
+++ material-components-android-master/lib/java/com/google/android/material/badge/BadgeState.java
@@ -207,7 +207,7 @@ public final class BadgeState {
             : storedState.badgeWithTextShapeAppearanceOverlayResId;
 
     currentState.backgroundColor =
-        storedState.backgroundColor == null
+            (storedState.backgroundColor == null || storedState.notSaveColor)
             ? readColorFromAttributes(context, a, R.styleable.Badge_backgroundColor)
             : storedState.backgroundColor;
 
@@ -220,7 +220,7 @@ public final class BadgeState {
 
     // Only set the badge text color if this attribute has explicitly been set, otherwise use the
     // text color specified in the TextAppearance.
-    if (storedState.badgeTextColor != null) {
+    if (storedState.badgeTextColor != null && !storedState.notSaveColor) {
       currentState.badgeTextColor = storedState.badgeTextColor;
     } else if (a.hasValue(R.styleable.Badge_badgeTextColor)) {
       currentState.badgeTextColor =
@@ -612,6 +612,15 @@ public final class BadgeState {
     currentState.numberLocale = locale;
   }
 
+  boolean isNotSaveColor() {
+    return currentState.notSaveColor;
+  }
+
+  void setNotSaveColor(boolean notSaveColor) {
+    overridingState.notSaveColor = notSaveColor;
+    currentState.notSaveColor = notSaveColor;
+  }
+
   /** Deprecated; badges now adjust to within bounds of first ancestor that clips its children */
   @Deprecated
   boolean isAutoAdjustedToGrandparentBounds() {
@@ -659,6 +668,7 @@ public final class BadgeState {
     private int maxCharacterCount = NOT_SET;
     private int maxNumber = NOT_SET;
     private Locale numberLocale;
+    private Boolean notSaveColor = false;
 
     @Nullable private CharSequence contentDescriptionForText;
     @Nullable private CharSequence contentDescriptionNumberless;
@@ -730,6 +740,7 @@ public final class BadgeState {
       numberLocale = (Locale) in.readSerializable();
       autoAdjustToWithinGrandparentBounds = (Boolean) in.readSerializable();
       badgeFixedEdge = (Integer) in.readSerializable();
+      notSaveColor = (Boolean) in.readSerializable();
     }
 
     public static final Creator<State> CREATOR =
@@ -786,6 +797,7 @@ public final class BadgeState {
       dest.writeSerializable(numberLocale);
       dest.writeSerializable(autoAdjustToWithinGrandparentBounds);
       dest.writeSerializable(badgeFixedEdge);
+      dest.writeSerializable(notSaveColor);
     }
   }
 }
Index: material-components-android-master/lib/java/com/google/android/material/badge/ExperimentalBadgeUtils.java
===================================================================
--- material-components-android-master.orig/lib/java/com/google/android/material/badge/ExperimentalBadgeUtils.java
+++ material-components-android-master/lib/java/com/google/android/material/badge/ExperimentalBadgeUtils.java
@@ -16,10 +16,8 @@
 
 package com.google.android.material.badge;
 
-import static androidx.annotation.experimental.Experimental.Level.ERROR;
 import static java.lang.annotation.RetentionPolicy.CLASS;
 
-import androidx.annotation.experimental.Experimental;
 import java.lang.annotation.ElementType;
 import java.lang.annotation.Retention;
 import java.lang.annotation.Target;
@@ -33,5 +31,4 @@ import java.lang.annotation.Target;
   ElementType.FIELD,
   ElementType.PACKAGE
 })
-@Experimental(level = ERROR)
 public @interface ExperimentalBadgeUtils {}
Index: material-components-android-master/lib/java/com/google/android/material/bottomnavigation/res/values/styles.xml
===================================================================
--- material-components-android-master.orig/lib/java/com/google/android/material/bottomnavigation/res/values/styles.xml
+++ material-components-android-master/lib/java/com/google/android/material/bottomnavigation/res/values/styles.xml
@@ -66,7 +66,6 @@
     <item name="itemPaddingTop">@dimen/m3_bottom_nav_item_padding_top</item>
     <item name="itemPaddingBottom">@dimen/m3_bottom_nav_item_padding_bottom</item>
     <item name="android:minHeight">@dimen/m3_bottom_nav_min_height</item>
-    <item name="materialThemeOverlay">@style/ThemeOverlay.Material3.BottomNavigationView</item>
     <item name="shapeAppearance">@style/ShapeAppearance.M3.Comp.NavBar.Container.Shape</item>
     <item name="activeIndicatorLabelPadding">@dimen/m3_navigation_item_active_indicator_label_padding</item>
     <item name="iconLabelHorizontalSpacing">@dimen/m3_navigation_item_active_indicator_label_padding</item>
@@ -74,14 +73,6 @@
 
   <style name="Widget.Material3.BottomNavigationView" parent="Base.Widget.Material3.BottomNavigationView"/>
 
-  <style name="ThemeOverlay.Material3.BottomNavigationView" parent="">
-    <item name="badgeStyle">@style/Widget.Material3.BottomNavigation.Badge</item>
-  </style>
-
-  <style name="Widget.Material3.BottomNavigation.Badge" parent="Widget.Material3.Badge">
-    <item name="verticalOffsetWithText">@dimen/m3_nav_badge_with_text_vertical_offset</item>
-  </style>
-
   <style name="Widget.Material3.BottomNavigationView.ActiveIndicator" parent="">
     <item name="android:width">@dimen/m3_bottom_nav_item_active_indicator_width</item>
     <item name="android:height">@dimen/m3_bottom_nav_item_active_indicator_height</item>
Index: material-components-android-master/lib/java/com/google/android/material/bottomsheet/BottomSheetBehavior.java
===================================================================
--- material-components-android-master.orig/lib/java/com/google/android/material/bottomsheet/BottomSheetBehavior.java
+++ material-components-android-master/lib/java/com/google/android/material/bottomsheet/BottomSheetBehavior.java
@@ -75,6 +75,8 @@ import androidx.core.view.insets.Protect
 import androidx.customview.view.AbsSavedState;
 import androidx.customview.widget.ViewDragHelper;
 import com.google.android.material.color.MaterialColors;
+import androidx.recyclerview.widget.RecyclerView;
+import androidx.viewpager2.widget.ViewPager2;
 import com.google.android.material.internal.ViewUtils;
 import com.google.android.material.internal.ViewUtils.RelativePadding;
 import com.google.android.material.motion.MaterialBackHandler;
@@ -1859,6 +1861,19 @@ public class BottomSheetBehavior<V exten
   }
 
   @Nullable
+  private View getCurrentViewFromViewPager2(@NonNull ViewPager2 viewPager) {
+    int currentItem = viewPager.getCurrentItem();
+    View child = viewPager.getChildAt(0);
+    if (child instanceof RecyclerView) {
+      if (((RecyclerView) child).getLayoutManager() == null)
+        return null;
+      child = ((RecyclerView) child).getLayoutManager().findViewByPosition(currentItem);
+    }
+
+    return child;
+  }
+
+  @Nullable
   @VisibleForTesting
   View findScrollingChild(View view) {
     if (view.getVisibility() != View.VISIBLE) {
@@ -1867,6 +1882,12 @@ public class BottomSheetBehavior<V exten
     if (view.isNestedScrollingEnabled()) {
       return view;
     }
+    if (view instanceof ViewPager2) {
+      View currentViewPagerChild = getCurrentViewFromViewPager2((ViewPager2) view);
+      if (currentViewPagerChild != null) {
+        return findScrollingChild(currentViewPagerChild);
+      }
+    }
     if (view instanceof ViewGroup) {
       ViewGroup group = (ViewGroup) view;
       for (int i = 0, count = group.getChildCount(); i < count; i++) {
Index: material-components-android-master/lib/java/com/google/android/material/card/MaterialCardViewHelper.java
===================================================================
--- material-components-android-master.orig/lib/java/com/google/android/material/card/MaterialCardViewHelper.java
+++ material-components-android-master/lib/java/com/google/android/material/card/MaterialCardViewHelper.java
@@ -167,7 +167,7 @@ class MaterialCardViewHelper {
     foregroundContentDrawable = new MaterialShapeDrawable();
     setShapeAppearance(shapeAppearanceModelBuilder.build());
 
-    iconFadeAnimInterpolator =
+    iconFadeAnimInterpolator = card.isInEditMode() ? AnimationUtils.LINEAR_INTERPOLATOR :
         MotionUtils.resolveThemeInterpolator(
             materialCardView.getContext(),
             R.attr.motionEasingLinearInterpolator,
@@ -231,7 +231,7 @@ class MaterialCardViewHelper {
 
     materialCardView.setBackgroundInternal(insetDrawable(bgDrawable));
     fgDrawable =
-        shouldUseClickableForeground() ? getClickableForeground() : foregroundContentDrawable;
+        shouldUseClickableForeground() ? getClickableForeground() : getForegroundContentDrawable();
     materialCardView.setForeground(insetDrawable(fgDrawable));
 
     // Card corner radius overrides the shape appearance in precedence.
@@ -253,6 +253,11 @@ class MaterialCardViewHelper {
     }
   }
 
+  @NonNull
+  private Drawable getForegroundContentDrawable() {
+    return foregroundContentDrawable;
+  }
+
   boolean isBackgroundOverwritten() {
     return isBackgroundOverwritten;
   }
@@ -328,7 +333,7 @@ class MaterialCardViewHelper {
   void updateClickable() {
     Drawable previousFgDrawable = fgDrawable;
     fgDrawable =
-        shouldUseClickableForeground() ? getClickableForeground() : foregroundContentDrawable;
+        shouldUseClickableForeground() ? getClickableForeground() : getForegroundContentDrawable();
     if (previousFgDrawable != fgDrawable) {
       updateInsetForeground(fgDrawable);
     }
Index: material-components-android-master/lib/java/com/google/android/material/materialswitch/res/values/dimens.xml
===================================================================
--- material-components-android-master.orig/lib/java/com/google/android/material/materialswitch/res/values/dimens.xml
+++ material-components-android-master/lib/java/com/google/android/material/materialswitch/res/values/dimens.xml
@@ -16,9 +16,9 @@
   -->
 
 <resources>
-  <dimen name="mtrl_switch_thumb_size">32dp</dimen>
+  <dimen name="mtrl_switch_thumb_size">24dp</dimen>
   <dimen name="mtrl_switch_thumb_icon_size">16dp</dimen>
-  <dimen name="mtrl_switch_track_width">@dimen/m3_comp_switch_track_width</dimen>
-  <dimen name="mtrl_switch_track_height">@dimen/m3_comp_switch_track_height</dimen>
+  <dimen name="mtrl_switch_track_width">44dp</dimen>
+  <dimen name="mtrl_switch_track_height">24dp</dimen>
   <dimen name="mtrl_switch_text_padding">16dp</dimen>
 </resources>
Index: material-components-android-master/lib/java/com/google/android/material/snackbar/Snackbar.java
===================================================================
--- material-components-android-master.orig/lib/java/com/google/android/material/snackbar/Snackbar.java
+++ material-components-android-master/lib/java/com/google/android/material/snackbar/Snackbar.java
@@ -283,7 +283,7 @@ public class Snackbar extends BaseTransi
   }
 
   @Nullable
-  private static ViewGroup findSuitableParent(View view) {
+  public static ViewGroup findSuitableParent(View view) {
     ViewGroup fallback = null;
     do {
       if (view instanceof CoordinatorLayout) {
@@ -471,6 +471,11 @@ public class Snackbar extends BaseTransi
         : userSetDuration;
   }
 
+  public Snackbar setOnClickListener(View.OnClickListener clickListener) {
+    getContentLayout().setOnClickListener(clickListener);
+    return this;
+  }
+
   /**
    * Sets the text color of the message specified in {@link #setText(CharSequence)} and {@link
    * #setText(int)}.
