Index: FFmpeg-n8.1.2/libavformat/hls.c
===================================================================
--- FFmpeg-n8.1.2.orig/libavformat/hls.c
+++ FFmpeg-n8.1.2/libavformat/hls.c
@@ -29,7 +29,6 @@
 
 #include "config_components.h"
 
-#include "libavformat/http.h"
 #include "libavutil/aes.h"
 #include "libavutil/avstring.h"
 #include "libavutil/avassert.h"
@@ -635,23 +634,6 @@ static int ensure_playlist(HLSContext *c
     return 0;
 }
 
-static int open_url_keepalive(AVFormatContext *s, AVIOContext **pb,
-                              const char *url, AVDictionary **options)
-{
-#if !CONFIG_HTTP_PROTOCOL
-    return AVERROR_PROTOCOL_NOT_FOUND;
-#else
-    int ret;
-    URLContext *uc = ffio_geturlcontext(*pb);
-    av_assert0(uc);
-    (*pb)->eof_reached = 0;
-    ret = ff_http_do_new_request2(uc, url, options);
-    if (ret < 0) {
-        ff_format_io_close(s, pb);
-    }
-    return ret;
-#endif
-}
 
 static int open_url(AVFormatContext *s, AVIOContext **pb, const char *url,
                     AVDictionary **opts, AVDictionary *opts2, int *is_http_out)
@@ -704,23 +686,9 @@ static int open_url(AVFormatContext *s,
     av_dict_copy(&tmp, *opts, 0);
     av_dict_copy(&tmp, opts2, 0);
 
-    if (is_http && c->http_persistent && *pb) {
-        ret = open_url_keepalive(c->ctx, pb, url, &tmp);
-        if (ret == AVERROR_EXIT) {
-            av_dict_free(&tmp);
-            return ret;
-        } else if (ret < 0) {
-            if (ret != AVERROR_EOF)
-                av_log(s, AV_LOG_WARNING,
-                    "keepalive request failed for '%s' with error: '%s' when opening url, retrying with new connection\n",
-                    url, av_err2str(ret));
-            av_dict_copy(&tmp, *opts, 0);
-            av_dict_copy(&tmp, opts2, 0);
-            ret = s->io_open(s, pb, url, AVIO_FLAG_READ, &tmp);
-        }
-    } else {
-        ret = s->io_open(s, pb, url, AVIO_FLAG_READ, &tmp);
-    }
+
+    ret = s->io_open(s, pb, url, AVIO_FLAG_READ, &tmp);
+
     if (ret >= 0) {
         // update cookies on http response with setcookies.
         char *new_cookies = NULL;
@@ -810,19 +778,6 @@ static int parse_playlist(HLSContext *c,
     int prev_n_segments = 0;
     int64_t prev_start_seq_no = -1;
 
-    if (is_http && !in && c->http_persistent && c->playlist_pb) {
-        in = c->playlist_pb;
-        ret = open_url_keepalive(c->ctx, &c->playlist_pb, url, NULL);
-        if (ret == AVERROR_EXIT) {
-            return ret;
-        } else if (ret < 0) {
-            if (ret != AVERROR_EOF)
-                av_log(c->ctx, AV_LOG_WARNING,
-                    "keepalive request failed for '%s' with error: '%s' when parsing playlist\n",
-                    url, av_err2str(ret));
-            in = NULL;
-        }
-    }
 
     if (!in) {
         AVDictionary *opts = NULL;
Index: FFmpeg-n8.1.2/libavformat/protocols.c
===================================================================
--- FFmpeg-n8.1.2.orig/libavformat/protocols.c
+++ FFmpeg-n8.1.2/libavformat/protocols.c
@@ -79,7 +79,11 @@ extern const URLProtocol ff_libzmq_proto
 extern const URLProtocol ff_ipfs_gateway_protocol;
 extern const URLProtocol ff_ipns_gateway_protocol;
 
-#include "libavformat/protocol_list.c"
+static const URLProtocol * const url_protocols[] = {
+    &ff_file_protocol,
+    &ff_https_protocol,
+    &ff_http_protocol,
+    NULL };
 
 const AVClass *ff_urlcontext_child_class_iterate(void **iter)
 {
